generators/chipyard/src/main/scala/upf/chisel/UPFAspect.scala

```diff
@@ -0,0 +1,31 @@
+// See LICENSE for license details
+package chipyard.upf.chisel
+
+import chisel3.experimental.{BaseModule}
+import chisel3.aop.Aspect
+import firrtl.{AnnotationSeq}
+import chipyard.TestHarness
+import freechips.rocketchip.diplomacy.LazyModule
+import firrtl.annotations.Annotation
+import chipyard.upf.firrtl.NameUPFAnnotation
+
+abstract class UPFAspect[T <: TestHarness](upf: UPFFunc.UPFFunction) extends Aspect[T] {
+
+  final override def toAnnotation(top: T): AnnotationSeq = {
+    upf(top.dut)
+    AnnotationSeq(ChiselUPFElement().getAnnotations()) // noop
```

Comment:
What happens if this returns no annotations (i.e. trying to remove the UPF annotation code)?

Reply:
Seems that overriding toAnnotation requires returning an AnnotationSeq (as defined in the Aspect class). However, returning an AnnotationSeq(Seq()) does seem to pass, so I will remove the ChiselUPFElement and Annotations code.